File: euler/haskell_solutions/two.hs
-- Project Euler Problem Number Two:
--
-- Find the sum of the even-valued terms of
-- the Fibonacci sequence below four million.

problemTwo :: Integer
problemTwo = sum $ filter (even) $ takeWhile (< 4000000) fib

-- The go-to Haskell Fibonacci function
fib = 0 : 1 : zipWith (+) fib (tail fib)

-- problemTwo ==> 4613732

File: euler/clojure_solutions/two.clj
; Project Euler Problem Number Two:
;
; Find the sum of the even-valued terms of
; the Fibonacci sequence below four million.

; Function for calculating Fibonacci numbers
(defn fib [a b]
  (lazy-cat [a] (fib b (+ a b))))

(defn problemTwo [n]
  (reduce + (for [f (fib 1 2) :when (zero? (mod f 2)) :while (< f n)] f)))

; (problemTwo 4000000) ==> 4613732

